### PR TITLE
fix(consider-interating-dictionary): fix false negatives

### LINKED ISSUE #5323: False negative for ``consider-iterating-dictionary`` when using ``not`` or ``list``
### Bug description

In the following sample `a.py`, only the last if raise a `consider-iterating-dictionary`

```python
metadata = {}
if "a" not in list(metadata.keys()):
    print(1)
if "a" not in metadata.keys():
    print(1)
if "a" in list(metadata.keys()):
    print(1)
if "a" in metadata.keys():
    print(1)
```    

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
************* Module a
a.py:8:10: C0201: Consider iterating the dictionary directly instead of calling .keys() (consider-iterating-dictionary)

------------------------------------------------------------------
Your code has been rated at 8.89/10 (previous run: 8.89/10, +0.00)
```

### Expected behavior

All four ifs are consider-iterating-dictionary

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.5
Python 3.8.10 (default, Sep 28 2021, 16:10:42) 
[GCC 9.3.0]
```

### PR DESCRIPTION
**Pull request by yushao2**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Added loop logic to check upwards (parents of node) until i hit a ``compare`` node, to circumvent the issue where `list(dict.keys())` results in the `node.parent` to be of type `call`

Also, added operator check for `not in`

Closes #5323

### CODE REVIEW

File: pylint/checkers/refactoring/recommendation_checker.py
Comment (DanielNoord) [inline pylint/checkers/refactoring/recommendation_checker.py:96]: Let's use something like this here:
https://github.com/PyCQA/astroid/blob/907818246c89ef938b23448e0cecc52b7c8693e0/astroid/nodes/node_classes.py#L4830-L4835

I think something like this should work:
```python
T_Node = TypeVar("T_Node", bound=NodeNG)

def get_if_node_type_ancestor(node: NodeNG, type: Tuple[Type[T_Node]]) -> Optional[T_Node]:
    """Return the first parent node that is any of the provided types (or None)"""
    for ancestor in node.node_ancestors():
        if isinstance(ancestor, type):
            return ancestor
    return None
```

Note that `type` is a `tuple` so we can use this function later to also check for multiple types.

We could at it to the checker utils file.

WIth this, we could do one check for ancestors and move all of the logic after that check

File: pylint/checkers/refactoring/recommendation_checker.py
Comment (DanielNoord) [inline pylint/checkers/refactoring/recommendation_checker.py:86]: I think it is better if we do the (performance-heavy) inference after we have done the check for ancestors (like we do in the current code).

With the changes in the other comment this should be easily doable.

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Thanks for this contribution!

We have currently closed `2.12` as we're trying to finalise that release, but this should definitely go in `2.13`. We will need to update the place of the changelog entry when we can (after release of `2.12`).

**Comment by Pierre-Sassoulas** [comment]
Nice to see  a new contribution from you @yushao2 :)

File: pylint/checkers/refactoring/recommendation_checker.py
Comment (yushao2) [reply pylint/checkers/refactoring/recommendation_checker.py]: really smart, i like that

File: pylint/checkers/refactoring/recommendation_checker.py
Comment (yushao2) [reply pylint/checkers/refactoring/recommendation_checker.py]: fixed this too

**Comment by yushao2** [comment]
> Nice to see a new contribution from you @yushao2 :)

great to be back, had a hard time adjusting to my new role as devops but slowly getting the hang of it (and thus more room to breathe and to contribute :D)

File: pylint/checkers/utils.py
Comment (DanielNoord) [inline pylint/checkers/utils.py:1692]: Could you move this to the top of the file? We tend to keep our `TypeVar`'s there.

File: pylint/checkers/refactoring/recommendation_checker.py
Comment (DanielNoord) [inline pylint/checkers/refactoring/recommendation_checker.py:93]:
```suggestion
                if op in {"in", "not in"}
```

File: pylint/checkers/refactoring/recommendation_checker.py
Comment (DanielNoord) [inline pylint/checkers/refactoring/recommendation_checker.py:94]: Just to check: is this ever untrue? I'm not sure how the `comparator` of `comp_ancestor.ops` works, but I can imagine this is always true?

File: pylint/checkers/refactoring/recommendation_checker.py
Comment (DanielNoord) [inline pylint/checkers/refactoring/recommendation_checker.py:86]:
```suggestion
        comp_ancestor = utils.get_node_first_ancestor_of_type(node, (nodes.Compare,))
```

File: pylint/checkers/refactoring/recommendation_checker.py
Comment (DanielNoord) [inline pylint/checkers/refactoring/recommendation_checker.py:89]:
```suggestion
            or comp_ancestor
```

The `isinstance` is unnecessary now that we use `utils.get_node_first_ancestor_of_type`.

**Review by DanielNoord — CHANGES_REQUESTED** [review]
This is really quite elegant! Thanks for the quick work 😄

File: pylint/checkers/refactoring/recommendation_checker.py
Comment (yushao2) [reply pylint/checkers/refactoring/recommendation_checker.py:93]: isn't tuple a better idea here?

```suggestion
                if op in ("in", "not in")
```

File: pylint/checkers/refactoring/recommendation_checker.py
Comment (yushao2) [reply pylint/checkers/refactoring/recommendation_checker.py:94]: it could be a comparison like 

`if a == {}.keys() in {}`

one of the tests fail if this check is omitted, see L68 of test file 

`VAR = [1, 2] == {}.keys() in {False}`

File: pylint/checkers/refactoring/recommendation_checker.py
Comment (Pierre-Sassoulas) [reply pylint/checkers/refactoring/recommendation_checker.py:93]: When checking for membership sets are faster, we have a check for that but it seems it's not activated right now.

**Review by DanielNoord — APPROVED** [review]
Looks good except for the set membership test. Thanks for this!

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Nice changes thank you @yushao2 ! I request change so we remember to move the changelog to 2.13 before merging.

**Review by DanielNoord — APPROVED** [review]

**Review by Pierre-Sassoulas — APPROVED** [review]
Small change, big result 👍

**Comment by DanielNoord** [comment]
@yushao2 Note that this PR was/is blocked by the release of `2.12.2` and the changelog entry in this PR is still in `2.12`. It should either be `2.12.2` or `2.13`. The other PR is also blocked!

**Comment by Pierre-Sassoulas** [comment]
Don't worry about this one, I'll fix the changelog when releasing 2.12.2

**Comment by yushao2** [comment]
That is my bad, i got confused. Sorry for the trouble

File: pylint/checkers/utils.py
Comment (cdce8p) [inline pylint/checkers/utils.py:1700]: Don't know if that has been considered. There is no need to limit `ancestor_type` to `Tuple[...]`. `isinstance` works fine with `Type[T_Node]` too. Ie.
```py
ancestor_type: Union[Type[T_Node], Tuple[Type[T_Node]]]
```
This has the advantage that now you can pass both a single type and a tuple of types to it.

File: pylint/checkers/refactoring/recommendation_checker.py
Comment (cdce8p) [inline pylint/checkers/refactoring/recommendation_checker.py:86]:
```py
comp_ancestor = utils.get_node_first_ancestor_of_type(node, nodes.Compare)
```
This could then be simplified.

Comment (DanielNoord) [reply pylint/checkers/refactoring/recommendation_checker.py:86]: Added to #5447 as a follow up issue to this.

I initially wanted to keep the typing of the function simple so thought it would be okay to only allow tuples, but we could allow both.

**Pull request merged**
